File: min_word_dist.py
# Time Complexity : O(n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


class Solution(object):
    def shortestDistance(self, wordsDict, word1, word2):
        """
        :type wordsDict: List[str]
        :type word1: str
        :type word2: str
        :rtype: int
        """
        min_dist = len(wordsDict) + 1

        # pointers
        i1 = None
        i2 = None

        for i, word in enumerate(wordsDict):
            if word == word1 or word == word2:

                if i2 is not None:
                    if word != i2:
                        min_dist = min(min_dist, i - i1)

                i2 = word
                i1 = i

        return min_dist

File: MinWordDist.java
//TC: O(nk) where n is the total words and k is the avg length of the word
//SC: O(1)
class Solution {
    public int shortestDistance(String[] wordsDict, String word1, String word2) {
        int p1 = -1;
        int p2 = -1;
        int min = Integer.MAX_VALUE;
        for(int i=0;i<wordsDict.length;i++){
            if(wordsDict[i].equals(word1))
                p1 = i;

            if(wordsDict[i].equals(word2))
                p2 = i;

            if(p1!=-1 && p2!=-1)
                min = Math.min(min,Math.abs(p1-p2));
        }
        return min;
    }
}